### PR TITLE
LPS-155882 Get the available readers and use them (logic based in javax.imageio.ImageIO.doWrite)

### PR DESCRIPTION
**Pull request by robertoDiaz**

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

**Comment by robertoDiaz** [comment]
ci:test:sf

**Comment by adolfopa** [comment]
ci:test:sf

**Comment by adolfopa** [comment]
ci:test:relevant

File: portal-impl/src/com/liferay/portal/image/ImageToolImpl.java [hunk 1]
```diff
@@ -704,25 +706,34 @@ public void write(
...
 
-		if (contentType.contains(TYPE_BMP)) {
-			ImageIO.write(renderedImage, "bmp", outputStream);
-		}
-		else if (contentType.contains(TYPE_GIF)) {
-			encodeGIF(renderedImage, outputStream);
-		}
-		else if (contentType.contains(TYPE_JPEG) ||
-				 contentType.contains("jpeg")) {
+		Iterator<ImageWriter> writerIterator =
```
Comment (adolfopa) [inline portal-impl/src/com/liferay/portal/image/ImageToolImpl.java:709]: This should be called either `iterator` or `imageWriterIterator`.

File: portal-impl/src/com/liferay/portal/image/ImageToolImpl.java [hunk 2]
```diff
@@ -704,25 +706,34 @@ public void write(
...
 
-			ImageIO.write(renderedImage, "jpeg", outputStream);
-		}
-		else if (contentType.contains(TYPE_PNG)) {
-			ImageIO.write(renderedImage, TYPE_PNG, outputStream);
-		}
-		else if (contentType.contains(TYPE_TIFF) ||
-				 contentType.contains("tif")) {
+		while (writerIterator.hasNext()) {
+			ImageWriter writer = writerIterator.next();
```
Comment (adolfopa) [inline portal-impl/src/com/liferay/portal/image/ImageToolImpl.java:713]: Can you rename `writer` to `imageWriter`?

File: portal-impl/src/com/liferay/portal/image/ImageToolImpl.java [hunk 3]
```diff
@@ -704,25 +706,34 @@ public void write(
...
+					_log.debug(exception);
+				}
+			}
+			finally {
+				writer.dispose();
+				imageOutputStream.flush();
+			}
 		}
+
+		throw new IOException();
```
Comment (adolfopa) [inline portal-impl/src/com/liferay/portal/image/ImageToolImpl.java:736]: Can you add an error message to the exception?

File: portal-impl/src/com/liferay/portal/image/ImageToolImpl.java [hunk 4]
```diff
@@ -704,25 +706,34 @@ public void write(
...
+				return;
+			}
+			catch (Exception exception) {
+				if (_log.isDebugEnabled()) {
+					_log.debug(exception);
+				}
+			}
+			finally {
+				writer.dispose();
+				imageOutputStream.flush();
```
Comment (adolfopa) [inline portal-impl/src/com/liferay/portal/image/ImageToolImpl.java:732]: There should be a newline between these two statements.

**Review by adolfopa — CHANGES_REQUESTED** [review]

**Comment by robertoDiaz** [comment]
Done 👍

**Comment by robertoDiaz** [comment]
Hey @liferay-lima there are some broken test, but check the test report is like a pain.

Could I ask @austinchiang for some help?

**Comment by austinchiang** [comment]
ci:reevaluate:1297401_3255

**Comment by austinchiang** [comment]
@robertoDiaz all the tests are failing due to an error in the console 
`java.lang.Exception:  ## ## 2 Liferay Exceptions were thrown ##  LIFERAY_ERROR: Unable to process file version 42499 LIFERAY_ERROR: Unable to register portal instance {"mvccVersion": 1, "companyId": 20098, "userId": 0, "userName": "", "createDate": "Mon Jun 13 09:35:59 GMT 2022", "modifiedDate": "Mon Jun 13 09:35:59 GMT 2022", "webId": "liferay.com", "mx": "liferay.com", "homeURL": "", "logoId": 0, "system": false, "maxUsers": 0, "active": true, "name": "Liferay", "legalName": "", "legalId": "
`

I don't think it's related since this has appeared in previous pulls as well (e.g. https://testray.liferay.com/home/-/testray/case_results/1877449104)

**Comment by austinchiang** [comment]
ci:test:relevant

**Comment by adolfopa** [comment]
ci:reevaluate:1297401_3299

**Comment by robertoDiaz** [comment]
I've found the cause of the log and I was able to reproduce it.

The problem is that in some cases the method does not receive the mime type but a simple type (for example "png" instead of "image/png".

I think that this is easy to fix.

**Comment by robertoDiaz** [comment]
ci:test:relevant

**Comment by adolfopa** [comment]
ci:test:relevant

**Pull request closed without merging**